### PR TITLE
Feature/wlts 29 security config

### PR DESCRIPTION
**Pull request by sem-hmcts**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/WLTS-29

### Change description ###
Remove IDAM token requirement on the `translation/cy` endpoint and the `PUT /dictionary` when the request is from the definition store

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/translate/security/filter/CustomPermitAllFilter.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    private boolean isPutDictionaryEndpoint(final HttpServletRequest request) {
+        return PUT_METHOD.equalsIgnoreCase(request.getMethod())
+            && ControllerConstants.DICTIONARY_URL.equalsIgnoreCase(request.getServletPath());
+    }
+
+    private boolean isPermittedService(final HttpServletRequest request) {
+        final String bearerToken = extractBearerToken(request);
+        final String serviceName = securityUtils.getServiceNameFromS2SToken(bearerToken);
```
Comment (sullivangareth) [inline src/main/java/uk/gov/hmcts/reform/translate/security/filter/CustomPermitAllFilter.java:58]: Can just pass in `request.getHeader(SERVICE_AUTHORIZATION)` to `getServiceNameFromS2SToken`, as that method removes the BEARER prefix from the token anyway....can then get rid of `extractBearerToken` method

Comment (sem-hmcts) [reply src/main/java/uk/gov/hmcts/reform/translate/security/filter/CustomPermitAllFilter.java]: Fixed in 2613f5856df83f8634ca449585342333779c169d

File: src/functionalTest/resources/features/F-002 - Put Dictionary/F-002.feature [hunk 2]
```diff
@@ -174,4 +174,45 @@ Feature: F-002: Put Dictionary Operation
...
+  @S-002.15 #WLTS-29 AC3
+  Scenario: Fail authentication and return 401 when IDAM token is missing and service is not definition store
+
+    When  a request is prepared with appropriate values
+    And the request [is missing IDAM AUTHORIZATION token header]
+    And the request [does not originate from ccd-definition]
+    And   the request [add a new English phrase]
+    And   it is submitted to call the [PUT dictionary] operation of [Translation Service]
+    Then  a negative response is received
+    And   the response [has the 401 Created status code]
```
[the remark this answers is not in the document]
Comment (sem-hmcts) [reply src/functionalTest/resources/features/F-002 - Put Dictionary/F-002.feature]: Fixed in 2613f5856df83f8634ca449585342333779c169d

File: src/main/java/uk/gov/hmcts/reform/translate/security/SecurityUtils.java [hunk 3]
```diff
@@ -57,22 +62,32 @@ public HttpHeaders authorizationHeaders() {
...
     public String getUserId() {
         return getUserInfo().getUid();
     }
 
     public String getUserToken() {
-        Jwt jwt = (Jwt) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
-        return jwt.getTokenValue();
+        return Optional.ofNullable(SecurityContextHolder.getContext().getAuthentication().getPrincipal())
+            .map(principal -> {
+                Jwt jwt = (Jwt) principal;
```
[the remark this answers is not in the document]
Comment (Thor-tech-of-metal) [reply src/main/java/uk/gov/hmcts/reform/translate/security/SecurityUtils.java:87]: If null principal is not an error ok cool . lets close it

File: src/main/java/uk/gov/hmcts/reform/translate/security/SecurityUtils.java [hunk 4]
```diff
@@ -57,22 +62,32 @@ public HttpHeaders authorizationHeaders() {
...
-        UserInfo userInfo = idamRepository.getUserInfo(getUserToken());
-        if (userInfo != null) {
-            log.info("SecurityUtils retrieved user info from idamRepository. User Id={}. Roles={}.",
-                     userInfo.getUid(),
-                     userInfo.getRoles());
-        }
-        return userInfo;
+        return Optional.ofNullable(getUserToken())
+            .map(userToken -> {
+                final UserInfo userInfo = idamRepository.getUserInfo(getUserToken());
```
[the remark this answers is not in the document]
Comment (sem-hmcts) [reply src/main/java/uk/gov/hmcts/reform/translate/security/SecurityUtils.java]: Modified in 114cf1147a0bf9d6e4fffe9bbaeac06f45c54b2e

**Review by sullivangareth — APPROVED** [review]

**Review by mattnayler — CHANGES_REQUESTED** [review]

File: src/functionalTest/java/uk/gov/hmcts/reform/translate/customvalue/GetTranslationsEvaluator.java [hunk 5]
```diff
@@ -16,7 +14,7 @@ public class GetTranslationsEvaluator implements CustomValueEvaluator {
 
     @Override
     public Boolean matches(CustomValueKey key) {
-        return CustomValueKey.S_003_1.equals(key);
+        return CustomValueKey.CONTAINS_TEST_TRANSLATIONS.equals(key);
```
[the remark this answers is not in the document]
Comment (sem-hmcts) [reply src/functionalTest/java/uk/gov/hmcts/reform/translate/customvalue/GetTranslationsEvaluator.java:17]: Done in 929e04beecd65ee8e70caa72f9b113f728a42513

File: src/test/java/uk/gov/hmcts/reform/translate/security/SecurityUtilsTest.java [hunk 6]
```diff
@@ -60,99 +76,190 @@ class SecurityUtilsTest {
...
+    @DisplayName("Assert user token and test roles absent scenarios")
+    class UserTokenAndAbsentRoles {
+        @BeforeEach
+        void prepare() {
+            doReturn(JWT).when(authentication).getPrincipal();
+            doReturn(authentication).when(securityContext).getAuthentication();
+        }
+
+        @Test
+        @DisplayName("Check manage-translations role not present")
```
[the remark this answers is not in the document]
Comment (sem-hmcts) [reply src/test/java/uk/gov/hmcts/reform/translate/security/SecurityUtilsTest.java]: This predates this ticket but done in 929e04beecd65ee8e70caa72f9b113f728a42513

File: src/test/java/uk/gov/hmcts/reform/translate/security/SecurityUtilsTest.java [hunk 7]
```diff
@@ -60,99 +76,190 @@ class SecurityUtilsTest {
...
+        void shouldReturnTrueWhenRoleNotPresent() {
+            assertFalse(underTest.hasRole("unknown-role"));
+        }
+
+        @Test
+        @DisplayName("Check any role not present")
+        void shouldReturnTrueWhenHasAnyOfTheseNotRolesPresent() {
```
[the remark this answers is not in the document]
Comment (sem-hmcts) [reply src/test/java/uk/gov/hmcts/reform/translate/security/SecurityUtilsTest.java]: This predates this ticket but done in 929e04beecd65ee8e70caa72f9b113f728a42513

File: src/functionalTest/resources/features/F-002 - Put Dictionary/F-002.feature [hunk 8]
```diff
@@ -174,4 +174,58 @@ Feature: F-002: Put Dictionary Operation
...
+    And   the request [add a new English phrase]
+    And   it is submitted to call the [PUT dictionary] operation of [Translation Service]
+    Then  a positive response is received
+    And   the response [has the 201 Created status code]
+    And   the response has all other details as expected
+    And   a successful call [to verify translations] as in [F-002_Verify]
+
+#-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+  @S-002.17
+  Scenario: Return 403 when an attempt to add English phrases to dictionary where a valid IDAM token is present and service is authorized
```
Comment (mattnayler) [inline src/functionalTest/resources/features/F-002 - Put Dictionary/F-002.feature:220]: Service is unauthorised?

**Review by mattnayler — APPROVED** [review]
Approved: with minor comment against FTA `S-002.17` discription

File: src/functionalTest/resources/features/F-002 - Put Dictionary/S-002.17.td.json [hunk 9]
```diff
@@ -0,0 +1,31 @@
+{
+  "_guid_": "S-002.17",
+  "_extends_": "Put_Dictionary_Base",
+  "title": "Return 403 when an attempt to add English phrases to dictionary where a valid IDAM token is present and service is authorized",
```
[the remark this answers is not in the document]
Comment (mattnayler) [reply src/functionalTest/resources/features/F-002 - Put Dictionary/S-002.17.td.json]: Resolved in dea53fa96a3f3f2369be7d7dcce65e90547e0e76

File: src/functionalTest/resources/features/F-002 - Put Dictionary/F-002.feature (hunk 8 above)
Comment (mattnayler) [reply src/functionalTest/resources/features/F-002 - Put Dictionary/F-002.feature]: Resolved in dea53fa96a3f3f2369be7d7dcce65e90547e0e76

**Review by sullivangareth — APPROVED** [review]

File: src/main/java/uk/gov/hmcts/reform/translate/security/filter/TranslateCyEndpointFilter.java [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+
+            SecurityContextHolder.getContext().setAuthentication(authenticationToken);
+            filterChain.doFilter(new HttpServletRequestWithoutAuthenticationHeader(request), response);
+            return;
+        }
+        filterChain.doFilter(request, response);
+    }
+
+    private boolean isPostTranslateCyEndpoint(final HttpServletRequest request) {
+        return POST_METHOD.equalsIgnoreCase(request.getMethod())
```
[the remark this answers is not in the document]
Comment (sem-hmcts) [reply src/main/java/uk/gov/hmcts/reform/translate/security/filter/TranslateCyEndpointFilter.java]: Done in 8c87ecb34cb485dfd4fe44e533a8d21235f936e8

**Pull request merged**